### ISSUE TITLE
Faster ODE solver + noise scheduling?

### ISSUE THREAD

**Issue by lcflint** [SWivid/F5-TTS#15]
Hey there, you guys have done an amazing job with this model!

I've been looking to use this model on edge devices. The performance is excellent, but it's a bit too slow for some of our deployment targets. Having tried the model, it looks like the biggest opportunity for performance improvements are with the ODE solver and reducing the number of required steps.

I've given a look at alternatives to `torchdiffeq` such as DPM-Solver, since it seems like that'd be able to reduce the number of steps needed to get good quality. However, I don't know how I'd implement this and it doesn't seem like the model is compatible with it for now.

There's also an explicit TODO in the code that says you guys will be implementing noise scheduling soon, which'd make the model compatible with DPM-Solver or another faster solver.

Do you plan on implementing a different solver soon to improve the performance? Or, do you have other plans to make the model faster?

**Comment by SWivid** [comment]
Thanks for using ~
1. The proposed Sway Sampling enables using less NFE and also a Euler ODE solver, while original E2 TTS using uniformly sampled flow step and the midpoint ODE solver. As far as I know, DPM-Solver enables less NFE steps but each step takes longer time (correct me if I'm wrong with it)
![image](https://github.com/user-attachments/assets/47989428-bd3a-4ea0-97eb-95ca20d9b861)
2. The training-time noise scheduler in Stable Diffusion 3 (which is also used by Seed-TTS) was speculated (by us) to help with that, say trained with such noise scheduler could enable less NFE steps during inference while maintaining same performance. Also, we have mentioned in our paper, rectification or distillation techniques may further help with inference efficiency.
3. If in your application scenario, there's no critical need for a multilingual ability, a smaller (e.g. 100M) single language model is fine and for sure faster. Also there's a lot we could do, e.g. parallel with classifier-free guidance inference (as it's two forward pass, one with condition, one without), or having a base model inference for first few steps (t->0), and apply a smaller one to do the following.

Maybe there's a lot I didn't cover, but I can say we're moving forward (though currently no plans to train for a separate solution😔).
Let's explore together ~

**Comment by lcflint** [comment]
Thanks for the prompt reply! Looks like there are plenty of avenues for performance gains.

It definitely sounds like parallelising the CFG evaluation would be a really good first step forward - since it doesn't seem to affect the structure of the model at all. I disabled the CFG evaluation during the forwards pass and found that it bought the performance down to acceptable levels.

From a high level it seems pretty simple. We'd evaluate the first pass for a given evaluation step, and while evaluating the next evaluation step's first pass we can evaluate the second pass for the step we previously did the first pass for at the same time. I'm just unsure how we'd be able to implement this with odeint.

Would be really interesting to look further into this, since any opportunity to optimise without having to retrain (esp. with the ~2x speedup parallelising these passes would likely give) is great.

**Comment by gaceladri** [comment]
I am also keen on exploring the potential for reducing the time to first token (TTFT) for real-time streaming TTS services. Based on my understanding, the time dependency of flow matching in sequential time steps, like those seen in CosyVoice, poses challenges for parallelization. Despite these constraints, TTFT in CosyVoice is noticeably slow, and I'm eager to test feasible optimization methods or alternative approaches you might recommend. Insights on potential solutions would be greatly appreciated.

**Comment by lcflint** [comment]
Pardon my previous (now deleted) comment, completely misinterpreted the way that the data dependencies of the model work

Just a follow up on the parallelization of the conditioned/non-conditioned passes, I tried to merge the two forwards passes into one call, but it only seems to have resulted in marginal performance gains (~0.75 to ~0.72RTF on my end)
```py
# Concatenate all of our vectors together for a combined prediction
combined_x = torch.cat([x, x], dim=0)
combined_cond = torch.cat([step_cond, torch.zeros_like(step_cond)], dim=0)
combined_text = torch.cat([text, torch.zeros_like(text)], dim=0)

# Predict our vector field components
combined_pred = self.transformer(
    x = combined_x, cond = combined_cond,
    text = combined_text, time = t,
    drop_audio_cond = False, drop_text = False
)
pred, null_pred = combined_pred.chunk(2, dim=0)
```
It seems like CUDA already recognised that the two passes had no dependencies on one another, so it's probably best to look in other directions
